public/app/core/utils/emitter.ts

```diff
@@ -37,7 +39,11 @@ export class Emitter {
    * Handles `event` with `handler()` when emitted.
    */
   on<T extends undefined>(event: AppEvent<T>, handler: () => void, scope?: any): void;
-  on<T extends Partial<T> extends T ? Partial<T> : never>(event: AppEvent<T>, handler: () => void, scope?: any): void;
+  on<T extends (U extends any ? Partial<T> : unknown) extends T ? Partial<T> : never, U = any>(
+    event: AppEvent<T>,
+    handler: () => void,
+    scope?: any
+  ): void;
```

Comment:
This is amazing typing :)

Reply:
I believe `T` always extends `Partial<T>`, so this could also have been simplified to `T extends Partial<T> extends T ? unknown : never`, which is probably more performant.  Since neither is particularly readable, may I also suggest adding a comment (I eventually concluded this was trying to check that `T` equals `Partial<T>`).